source-kafka/src/configuration.rs

```diff
@@ -0,0 +1,145 @@
+#![allow(dead_code)]
+
+use std::{fmt::Display, io::Read};
+
+use serde::{de::Visitor, Deserialize, Serialize};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("failed to read the configuration file")]
+    File(#[from] std::io::Error),
+
+    #[error("failed to parse the file as valid json")]
+    Parsing(#[from] serde_json::Error),
+
+    #[error("bootstrap servers are required to make the initial connection")]
+    NoBootstrapServersGiven,
+}
+
+#[derive(Deserialize, Default, Serialize)]
+pub struct Configuration {
+    pub bootstrap_servers: Vec<BootstrapServer>,
+}
+
+impl Configuration {
+    pub fn parse<R: Read>(reader: R) -> Result<Configuration, Error> {
+        let configuration: Configuration = serde_json::from_reader(reader)?;
+
+        if configuration.bootstrap_servers.is_empty() {
+            return Err(Error::NoBootstrapServersGiven);
+        }
+
+        Ok(configuration)
+    }
+
+    pub fn brokers(&self) -> String {
+        self.bootstrap_servers
+            .iter()
+            .map(|url| url.to_string())
+            .collect::<Vec<String>>()
+            .join(",")
+    }
+}
+
+/// The initial servers in the Kafka cluster to initially connect to. The Kafka
+/// client will be informed of the rest of the cluster nodes by connecting to
+/// one of these nodes.
+//
+// Note: The Rust `url` crate doesn't like `ip:port` pairs without a scheme.
+// Kafka doesn't specify a scheme for the `bootstrap_servers` values, so
+// expecting anyone to add one is really odd. Thus, we parse these values
+// ourselves.
+pub struct BootstrapServer {
+    host: String,
+    port: u16,
+}
+
+impl BootstrapServer {
+    pub fn new(host: impl Into<String>, port: u16) -> Self {
+        Self {
+            host: host.into(),
+            port,
+        }
+    }
+}
+
+impl Display for BootstrapServer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}:{}", &self.host, self.port)
+    }
+}
+
+impl Serialize for BootstrapServer {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for BootstrapServer {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(BootstrapServerVisitor)
+    }
+}
+
+struct BootstrapServerVisitor;
+
+impl<'de> Visitor<'de> for BootstrapServerVisitor {
```

Comment:
It looks like a `SocketAddr` is explicitly a pair with an IP Address and a Port. This means that `ToSocketAddrs` is going to attempt to resolve the DNS if you give it a hostname.

```rust
impl ToSocketAddrs for (&str, u16) {
    fn to_socket_addrs(&self) -> io::Result<vec::IntoIter<SocketAddr>> {
        let (host, port) = *self;
        // try to parse the host as a regular IP address first
        if let Ok(addr) = host.parse::<Ipv4Addr>() {
            let addr = SocketAddrV4::new(addr, port);
            return Ok(vec![SocketAddr::V4(addr)].into_iter());
        }
        if let Ok(addr) = host.parse::<Ipv6Addr>() {
            let addr = SocketAddrV6::new(addr, port, 0, 0);
            return Ok(vec![SocketAddr::V6(addr)].into_iter());
        }
        resolve_socket_addr((host, port).try_into()?) // <------------
    }
}
```
https://doc.rust-lang.org/stable/src/std/net/addr.rs.html#967

Is this okay? :thinking: Thinking it through, if we invoke `check`/`discover`/`read`, we're going to try and connect to this broker. If we try to connect to the broker, we'll be resolving the DNS hostname. If we resolve it during configuration parsing and it fails, we wouldn't be able to resolve it 3 seconds later when we try to actually connect either. This makes me think it's alright to eagerly resolve this host:port pair. Seem sound to you?

Reply:
I _think_ that's probably fine, but my gut instinct here is to err on the side of safety and let the kafka library handle DNS resolution. Otherwise there's potential that we do the resolution differently than rdkafka in a way that might make a difference.

Reply:
Yeah, I think keeping configuration parsing as a stateless activity is probably worth the boilerplate of parsing the `host:port` pairs. :+1: